### PR TITLE
Serde Resolver

### PR DESCRIPTION
**Pull request by 99-not-out**
Adds a "serve resolver" which given a serdes config with `:key-serde`  and `:value-serde` keys in it, replaces the values for those keys with the actual  implementations for the required serde. Accepts options to specify any extra information which may be needed to create a serde, e.g. the avro serdes needs to know the  schema registry URL.

The new metadata format looks like:
```
(def +topic-metadata+
  {"input"
   {:topic-name "input"
    :partition-count 1
    :replication-factor 1
    :key-serde {:serde-qualified-keyword :jackdaw.serdes/string-serde}
    :value-serde {:serde-qualified-keyword :jackdaw.serdes.avro.confluent/serde
                  :schema-filename "value-schema.json"
                  :key? false}}

   "output"
   {:topic-name "output"
    :partition-count 1
    :replication-factor 1
    :key-serde {:serde-qualified-keyword :jackdaw.serdes/string-serde}
    :value-serde {:serde-qualified-keyword :jackdaw.serdes.avro.confluent/serde
                  :schema-filename "value-schema.json"
                  :key? false}}})
```

The resolver namespace contains a function called `serde-resolver` that returns a function of arity one which takes a **serde config** (as opposed to a topic config) and resolves the serde. Only one serde is resolved at a time.

This is the serde spec:
```
(s/def ::serde-qualified-keyword qualified-keyword?)
(s/def ::schema string?)
(s/def ::schema-filename string?)
(s/def ::key? boolean?)

(s/def ::jackdaw.serde/serde
  (s/keys :req-un [::serde-qualified-keyword]
          :opt-un [::schema
                   ::schema-filename
                   ::key?]))
```

To update some topic "metadata" with serdes resolved, there is still a bit of glue code (but not very much!):
```
(def serde-resolver
  (partial resolver/serde-resolver :schema-registry-url "http://localhost:8081"))

(def topic-metadata
  (memoize (fn []
             (fmap #(assoc % :key-serde ((serde-resolver) (:key-serde %))
                           :value-serde ((serde-resolver) (:value-serde %)))
                   +topic-metadata+))))
```

There are some examples added also in this PR.

### CODE REVIEW

File: src/jackdaw/serdes/resolver.clj [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      "{\"type\":\"string\"}"
+      (when-let [r (.getResource
+                     (.getContextClassLoader (Thread/currentThread))
+                     (str (:value-schema-name topic-config) ".json"))]
+        (slurp r)))))
+
+(defn create-avro-serde [schema-registry-url schema-loader topic-name key?]
+  (avro-serde/serde
+    ;; Registry of avro <-> clj types
+    (merge avro-serde/+base-schema-type-registry+ avro-serde/+UUID-type-registry+)
```
Comment (99-not-out) [inline src/jackdaw/serdes/resolver.clj]: Would need to be passed in

File: src/jackdaw/serdes/resolver.clj [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  {:avro-key (fn [topic-config]
+               (create-avro-serde
+                 schema-registry-url schema-loader topic-config true))
+   :avro-value (fn [topic-config]
+                 (create-avro-serde
+                   schema-registry-url schema-loader topic-config false))
+   :json (fn [_]
+           (json-serde/serde))
+   :string (fn [_]
+             (org.apache.kafka.common.serialization.Serdes/String))})
```
Comment (99-not-out) [inline src/jackdaw/serdes/resolver.clj]: Add EDN, number etc. ...

File: src/jackdaw/serdes/resolver.clj [hunk 3]
```diff
@@ -0,0 +1,50 @@
+(ns jackdaw.serdes.resolver
+  (:require
+    [jackdaw.serdes.avro :as avro-serde]
+    [jackdaw.serdes.json :as json-serde]))
+
+(defn local-schema-loader []
+  (fn [topic-config key?]
+    (if key?
+      "{\"type\":\"string\"}"
```
Comment (99-not-out) [inline src/jackdaw/serdes/resolver.clj]: Allow to be altered

File: src/jackdaw/serdes/resolver.clj [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+      (when-let [r (.getResource
+                     (.getContextClassLoader (Thread/currentThread))
+                     (str (:value-schema-name topic-config) ".json"))]
```
Comment (99-not-out) [inline src/jackdaw/serdes/resolver.clj]: Support searching multiple resource locations (and make configurable)

File: src/jackdaw/serdes/resolver.clj [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                        (create-serdes-lookup schema-registry-url schema-loader)
+                        (get topic-config k))]
+                (f topic-config)))]
+      (merge
+        topic-config
+        {:key-serde (get-serde :key-serde)
+         :value-serde (get-serde :value-serde)}))))
+
+(defn local-serdes-resolver [schema-registry-url]
+  (serdes-resolver schema-registry-url (local-schema-loader)))
```
Comment (99-not-out) [inline src/jackdaw/serdes/resolver.clj]: Add a `remote-series-resolver` which takes the schema to use form the registry rather than local disk.

**Comment by creese** [comment]
https://github.com/FundingCircle/jackdaw/pull/71 is an alternative. Instead of introducing helpers, we show by example.

**Comment by creese** [comment]
I've tweaked this some.

The new metadata format looks like:
```
(def +topic-metadata+
  {"input"
   {:topic-name "input"
    :partition-count 1
    :replication-factor 1
    :key-serde {:serde-qualified-keyword :jackdaw.serdes/string-serde}
    :value-serde {:serde-qualified-keyword :jackdaw.serdes.avro.confluent/serde
                  :schema-filename "value-schema.json"
                  :key? false}}

   "output"
   {:topic-name "output"
    :partition-count 1
    :replication-factor 1
    :key-serde {:serde-qualified-keyword :jackdaw.serdes/string-serde}
    :value-serde {:serde-qualified-keyword :jackdaw.serdes.avro.confluent/serde
                  :schema-filename "value-schema.json"
                  :key? false}}})
```

The resolver namespace now contains a function called `serde-resolver` that returns a function of arity one which takes a **serde config** (as opposed to a topic config) and resolves the serde. This is the key difference: only one serde is resolved.

This is the serde spec:
```
(s/def ::serde-qualified-keyword qualified-keyword?)
(s/def ::schema string?)
(s/def ::schema-filename string?)
(s/def ::key? boolean?)

(s/def ::jackdaw.serde/serde
  (s/keys :req-un [::serde-qualified-keyword]
          :opt-un [::schema
                   ::schema-filename
                   ::key?]))
```

To get the topic metadata with serdes resolved, we need a bit of glue code (not very much!):
```
(def serde-resolver
  (partial resolver/serde-resolver :schema-registry-url "http://localhost:8081"))

(def topic-metadata
  (memoize (fn []
             (fmap #(assoc % :key-serde ((serde-resolver) (:key-serde %))
                           :value-serde ((serde-resolver) (:value-serde %)))
                   +topic-metadata+))))
```

Turns out, including the serde resolver in Jackdaw is a Very Good Idea as resolving one serde is nearly all of the work!

I've added the examples from https://github.com/FundingCircle/jackdaw/pull/73 to this PR. Have a look.

**Comment by 99-not-out** [comment]
Thanks for the changes Charles, I've got the tests added back and everything working. LGTM now.

File: examples/serdes/dev/system.clj [hunk 6]
```diff
@@ -0,0 +1,67 @@
+(ns system
+  "Functions to start and stop the system, used for interactive
+  development.
+
+  The `system/start` and `system/stop` functions are required by the
+  `user` namespace and should not be called directly."
+  (:require [clojure.string :as str]
+            [clojure.tools.logging :as log]
```
Comment (AndreaCrotti) [inline examples/serdes/dev/system.clj]: This file is a very good candidate for integrant and integrant.repl

https://github.com/weavejester/integrant/
https://github.com/weavejester/integrant-repl

specially if we leave it as dev only dependency

**Comment by creese** [comment]
@99-not-out I couldn't find a commit with your changes to `jackdaw.serdes.resolver-test`. Did you maybe force push?

**Comment by 99-not-out** [comment]
Yeah - I squashed it all.
Which I now also realise destroyed the record of your contributions ...

File: examples/serdes/dev/system.clj (hunk 6 above)
Comment (kidpollo) [reply examples/serdes/dev/system.clj]: Yes but its not needed for an example IMO. I am generally dependency adverse :P

Comment (creese) [reply examples/serdes/dev/system.clj]: I'd be fine with integrant if it DRYs out the code but not for this PR.

File: src/jackdaw/serdes.clj [hunk 7]
```diff
@@ -0,0 +1,8 @@
+(ns jackdaw.serdes
+  "Implements basic SerDes (Serializer/Deserializer)."
```
Comment (kidpollo) [inline src/jackdaw/serdes.clj]: There are other basic serdes to add here right? Along with no so basic ones like `uuid`

**Comment by kidpollo** [comment]
To clarify. If I have a custom serde to use it I would just need to `{:serde-qualified-keyword :ns.path.to/my-serde-fn}` ?

File: src/jackdaw/serdes.clj (hunk 7 above)
Comment (creese) [reply src/jackdaw/serdes.clj]: Most like. After this is merged, feel free to any you find are missing.

**Comment by creese** [comment]
`:jackdaw.specs/serde-qualified-keyword` is now `:jackdaw.specs/serde-keyword`

**Pull request merged**